### PR TITLE
PersonFacadeEjb.getIndexList: read Ids first #8938

### LINKED ISSUE #8938: Improve performance of PersonFacadeEjb.getIndexList [5]
### Problem Description

This summarizes the problem for getIndexList methods in general, not only for Persons.

For large datasets, some of the `getIndexList` methods are slow, performance should be improved here.

Dataset:

- 1225343 persons
- 85677 cases
- 114782 contacts
- 236583 tasks
- 2600890 immunizations
- 1991 vaccinations
- 15693 samples
- 11846 events
- 534 eventparticipants

Result of a performance analysis:
image

The runtime is mainly due to long running SQL queries, optimization is needed here.

Some preliminary insights:
- in the following query from `PersonFacadeEjb.getIndexList`, runtime is reduced by facto 0.5 when omitting the joins on district, subselect and `case` statement. Since the query is limited to 100 result, a viable solution might be to select Ids /Uuids first and fill the remaining data after for the returned entities. This, however, with respect to the properties possibly being used a sorting criteria.

<details>

```
    select
        distinct person0_.uuid as col_0_0_,
        person0_.firstName as col_1_0_,
        person0_.lastName as col_2_0_,
        person0_.approximateAge as col_3_0_,
        person0_.approximateAgeType as col_4_0_,
        person0_.birthdate_dd as col_5_0_,
        person0_.birthdate_mm as col_6_0_,
        person0_.birthdate_yyyy as col_7_0_,
        person0_.sex as col_8_0_,
        district2_.name as col_9_0_,
        location1_.street as col_10_0_,
        location1_.houseNumber as col_11_0_,
        location1_.postalCode as col_12_0_,
        location1_.city as col_13_0_,
        (select
            personcont12_.contactInformation 
        from
            PersonContactDetail personcont12_ 
        where
            personcont12_.person_id=person0_.id 
            and personcont12_.primaryContact=true 
            and personcont12_.personContactDetailType='PHONE') as col_14_0_,
        (select
            personcont13_.contactInformation 
        from
            PersonContactDetail personcont13_ 
        where
            personcont13_.person_id=person0_.id 
            and personcont13_.primaryContact=true 
            and personcont13_.personContactDetailType='EMAIL') as col_15_0_,
        person0_.changeDate as col_16_0_,
        case 
            when 1=1 
            or (cases5_.reportingUser_id is not null) 
            and cases5_.reportingUser_id=8113961 
            or 1=1 
            or (contacts6_.reportingUser_id is not null) 
            and contacts6_.reportingUser_id=8113961 
            or 1=1 
            or 1=1 
            or (eventparti8_.reportingUser_id is not null) 
            and eventparti8_.reportingUser_id=8113961 
            or 1=1 
            or (travelentr9_.reportingUser_id is not null) 
            and travelentr9_.reportingUser_id=8113961 
            or 1=1 
            or (immunizati11_.reportingUser_id is not null) 
            and immunizati11_.reportingUser_id=8113961 
            or 1=1 then true 
            else false 
        end as col_17_0_
    from
        Person person0_ 
    left outer join
        Location location1_ 
            on person0_.address_id=location1_.id 
    left outer join
        District district2_ 
            on location1_.district_id=district2_.id 
    left outer join
        Region region3_ 
            on location1_.region_id=region3_.id 
    left outer join
        Community community4_ 
            on location1_.community_id=community4_.id 
    inner join
        cases cases5_ 
            on person0_.id=cases5_.person_id 
    left outer join
        contact contacts6_ 
            on person0_.id=contacts6_.person_id 
    left outer join
        cases case7_ 
            on contacts6_.caze_id=case7_.id 
    left outer join
        EventParticipant eventparti8_ 
            on person0_.id=eventparti8_.person_id 
    left outer join
        travelentry travelentr9_ 
            on person0_.id=travelentr9_.person_id 
    left outer join
        cases case10_ 
            on travelentr9_.resultingCase_id=case10_.id 
    left outer join
        immunization immunizati11_ 
            on person0_.id=immunizati11_.person_id 
    where
        cases5_.deleted=false 
    order by
        person0_.changeDate desc limit 100;
```

</details>

- for `DirectoryImmunizationService.getIndexList` a similarly drastic reduction of runtimes can be observed by ommiting joins (e.g., district)
- it is notable that `getIndexList` methods are performing reasonably for `CaseFacadeEjb` and `ContactFacadeEjb`, which deserves a closer look when working on `PersonFacadeEjb.getIndexList`, `TaskFacadeEjb.getIndexList`, `DirectoryImmunizationService.getIndexList` and the like
- adding database indices in relevant places should be considered

### Proposed Change

We will start by improving the performance of the PersonFacadeEjb.getIndexList method. If that works out, we should create separate issues to improve the performance for cases, contacts, events, event participants, samples, tasks, immunizations and travel entries (so basically for every entity that has a directory view).

#8946 has proven that by splitting the query into fetching the sorted ids first and afterwards doing another query to get the needed data, we can improve the performance a lot, because the database does not have to fetch and sort the whole table to then only return a limited number of results.

Before starting this make sure to get a performance test database and

- [ ] Create an explain analyze of the current query execute by getIndexList following [this guide](https://github.com/hzi-braunschweig/SORMAS-Project/blob/b494beb7e245f65885a7a4187f8c66323a24c0a8/docs/TROUBLESHOOTING.md#run-analysis-of-a-sql-query-sormas-docker) (you can also get the SQL by enabling the output in the persistence.xml). The annoying part here is to replace the params

The current implementation then should be changed as follows:

- [ ] 1. Rewrite PersonFacadeEjb.getIndexList to first fetch the needed ids (see pattern in PersonService.getAllAfter), then fetch the IndexDto by id with IN-clause. https://github.com/hzi-braunschweig/SORMAS-Project/issues/8946
- [ ] 2. Add indices for sorting by first name and by last name. Adding indices for sorting by other fields probably doesn't make a lot of sense.

Finally:

- [ ] Measure the query performance again (this time it will obviously be two queries). 
- [ ] If this works out as expected get in contact with @StefanKock or @MartinWahnschaffe to refine #8938 for the other entities.

### Acceptance Criteria

* PersonFacadeEjb.getIndexList returns the same results as before
* The performance of the method is improved

### PR DESCRIPTION
**Pull request by syntakker**
Fixes #8938

### CODE REVIEW

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java [hunk 1]
```diff
@@ -1412,6 +1405,84 @@ public List<PersonIndexDto> getIndexList(PersonCriteria criteria, Integer first,
...
+
+		Predicate filter = createIndexListFilter(criteria, personQueryContext);
+		if (filter != null) {
+			cq.where(filter);
+		}
+		cq.distinct(true);
+
+		List<Tuple> persons = QueryHelper.getResultList(em, cq, first, max);
+
+		logger.debug(
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1487]: Maybe reduce to log level trace?

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java [hunk 2]
```diff
@@ -1349,7 +1349,13 @@ public List<PersonIndexDto> getIndexList(PersonCriteria criteria, Integer first,
 			person.get(Person.CHANGE_DATE),
 			JurisdictionHelper.booleanSelector(cb, service.inJurisdictionOrOwned(personQueryContext)));
 
-		cq.where(person.get(Person.ID).in(indexListIds));
+		Predicate filter = person.get(Person.ID).in(indexListIds);
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1352]: Why is this needed? This indexListFilter is already part of the getIndexListIds logic

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java [hunk 3]
```diff
@@ -1412,6 +1405,84 @@ public List<PersonIndexDto> getIndexList(PersonCriteria criteria, Integer first,
...
+		final Root<Person> person = cq.from(Person.class);
+
+		final PersonQueryContext personQueryContext = new PersonQueryContext(cb, cq, person);
+		final PersonJoins personJoins = personQueryContext.getJoins();
+		personJoins.configure(criteria);
+
+		List<Selection<?>> selects = new ArrayList<>();
+		selects.add(person.get(Person.ID));
+
+		if (sortProperties != null && sortProperties.size() > 0) {
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1431]: This `sortproperties` block is completely duplicated and should by extracted into a separate method.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java [hunk 4]
```diff
@@ -1391,7 +1391,7 @@ public List<PersonIndexDto> getIndexList(PersonCriteria criteria, Integer first,
 			cq.orderBy(cb.desc(person.get(Person.CHANGE_DATE)));
 		}
 
-		List<PersonIndexDto> persons = QueryHelper.getResultList(em, cq, first, max);
+		List<PersonIndexDto> persons = QueryHelper.getResultList(em, cq, null, null);
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1400]: If we do not use `first` and `max` (because it's applied to the ids query), we should not call `QueryHelper` but use `em.createQuery(cq).getResultList()` directly.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java [hunk 5]
```diff
@@ -1350,10 +1346,7 @@ public List<PersonIndexDto> getIndexList(PersonCriteria criteria, Integer first,
 			person.get(Person.CHANGE_DATE),
 			JurisdictionHelper.booleanSelector(cb, service.inJurisdictionOrOwned(personQueryContext)));
 
-		Predicate filter = createIndexListFilter(criteria, personQueryContext);
-		if (filter != null) {
-			cq.where(filter);
-		}
+		cq.where(person.get(Person.ID).in(indexListIds));
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java]: The in clause could surpass the parameter limit, so please encapsulate the data fetch query with `IterableHelper.executeBatched` It also deals with empty id list so no additional empty check is then needed.

**Review by StefanKock — CHANGES_REQUESTED** [review]
The approach looks promising. I have some remarks concerning the code.
Additionally, please place the ticket id in front in the commit message.

I suggest to squash all changes (also with my remarked changes) into a new commit and to push force it on this branch.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java (hunk 1 above)
Comment (syntakker) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1487]: Done.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java (hunk 3 above)
Comment (syntakker) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1431]: Done.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java (hunk 4 above)
Comment (syntakker) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1400]: Done.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java (hunk 2 above)
Comment (syntakker) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java]: `indexListFilter` has been reintroduced in commit 421860c after failure of test `PersonFacadeEjbUserFilterTest.testGetPersonIndexListWhenSeveralAssociations`, where a person entry is duplicated with different values for `PersonIndexDto.isInJurisdiction` (see https://github.com/hzi-braunschweig/SORMAS-Project/actions/runs/3429250031/jobs/5714616024). In this case, one of these two distinct entries is filtered out by the `indexListFilter`.

**Comment by syntakker** [comment]
For displaying the person index page in the Vaadin UI, a relevant contribution to execution time originates in the queries for the person count per association.

Execution time per association:
- by immunization: 4320.867 ms
- by contact: 840.270 ms
- by case: 299.102 ms

Queries using `explain` and analysis results can be found here:
[8938_personCounts.zip](https://github.com/hzi-braunschweig/SORMAS-Project/files/10016206/8938_personCounts.zip)

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java [hunk 6]
```diff
@@ -1425,19 +1390,43 @@ private List<Long> getIndexListIds(PersonCriteria criteria, Integer first, Integ
...
-			List<Order> order = new ArrayList<Order>(sortProperties.size());
+		logger.trace(
+			"getIndexListIds() finished. association={}, count={}, {}ms",
+			Optional.ofNullable(criteria).orElse(new PersonCriteria()).getPersonAssociation().name(),
+			persons.size(),
+			DateHelper.durationMillies(startTime));
+		return persons.stream().map(t -> t.get(0, Long.class)).collect(Collectors.toList());
+	}
+
+	private List<Selection<?>> addSortProperties(List<SortProperty> sortProperties, PersonQueryContext personQueryContext) {
```
Comment (StefanKock) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java:1415]: Please rename to `sortBy` to point out that the logic does not only compile the sort selections, but also applies them.

**Review by StefanKock — CHANGES_REQUESTED** [review]
Only one finding on new method, the rest looks clean

File: sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java (hunk 6 above)
Comment (syntakker) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/person/PersonFacadeEjb.java]: Done.

**Review by StefanKock — APPROVED** [review]
Looks okay for me, but please squash merge it with this message:
"#8938 PersonFacade.getIndexList: read Ids first"
